### PR TITLE
feat(pr): expandCodeOwnersGroups

### PR DESCRIPTION
**Pull request by bbckr**
## Changes

Add a config option `expandCodeOwnersGroups` that when enabled alongside `assigneesFromCodeOwners`, will expand the code owners group into a list of members in place of the group.

Test MR here on real repository [here](https://gitlab.com/bckr/renovate-testing/-/merge_requests/3).

Only supported for Gitlab.

## Context

Closes #25300

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository (see: [Gitlab MR](https://gitlab.com/bckr/renovate-testing/-/merge_requests/3))

### CODE REVIEW

File: lib/config/options/index.ts
Comment (viceice) [inline lib/config/options/index.ts:2101]: add platform property with only gitlab, as this is only supported on gitlab

File: lib/modules/platform/gitlab/index.ts
Comment (viceice) [inline lib/modules/platform/gitlab/index.ts:1358]:
```suggestion
      expandedReviewersOrAssignees.push(...members);
```

File: lib/modules/platform/gitlab/index.ts
Comment (viceice) [inline lib/modules/platform/gitlab/index.ts:1360]: log a debug message

File: lib/workers/repository/update/pr/participants.ts
Comment (viceice) [inline lib/workers/repository/update/pr/participants.ts:15]: @rarkins Should we warn if not supported but enabled?

File: lib/config/options/index.ts
Comment (rarkins) [reply lib/config/options/index.ts:2101]: `supportPlatforms`

File: lib/workers/repository/update/pr/participants.ts
Comment (bbckr) [reply lib/workers/repository/update/pr/participants.ts:15]: Would be useful, updated to warn

**Review by rarkins — CHANGES_REQUESTED** [review]

**Comment by bbckr** [comment]
@rarkins @viceice Updated the PR with your suggested changes. I have not gotten around to testing it against a live GitLab test repository, but will take a look into that hopefully this weekend, if not Monday.

**Comment by viceice** [comment]
please use the re-request review button when you've done testing

File: docs/usage/configuration-options.md
Comment (rarkins) [inline docs/usage/configuration-options.md:1244]:
```suggestion
If configured, Renovate will expand any matching `CODEOWNERS` groups into a full list of group members and assign them individually instead of the group.
```

File: docs/usage/configuration-options.md
Comment (rarkins) [inline docs/usage/configuration-options.md:1242]:
```suggestion
## expandCodeOwnersGroups
```

For consistency with `assigneesFromCodeOwners`

File: lib/config/options/index.ts
Comment (rarkins) [inline lib/config/options/index.ts:2097]:
```suggestion
    name: 'expandCodeOwnersGroups',
```

**Review by rarkins — CHANGES_REQUESTED** [review]
Name change for consistency

**Review by rarkins — DISMISSED** [review]

File: lib/modules/platform/gitlab/index.ts
Comment (viceice) [inline lib/modules/platform/gitlab/index.ts:1359]: isn't there a better way to differentiate between user and group?

File: lib/workers/repository/update/pr/participants.ts
Comment (viceice) [inline lib/workers/repository/update/pr/participants.ts:15]: why remove the `@` symbol? shouldn't it be used to detect groups instead of doing failing API requests?

Comment (bbckr) [reply lib/workers/repository/update/pr/participants.ts:15]: Using the @ symbol is not a fool-proof way to differentiate users from groups, since while user emails don't require the beginning @, usernames do. Still, I can update the PR to filter the user emails from the list, but there will still be failing API requests where something like `@username` is used.

Additionally, removing the @ symbol is required when searching for a Group against the Gitlab API, so it needs to happen before the request is made.

File: lib/modules/platform/gitlab/index.ts
Comment (bbckr) [reply lib/modules/platform/gitlab/index.ts:1359]: See the below comment, but we can have user@email.com, @username, and @group. We can filter everything w/o an @ before, but will still need to see if the API request will fail for when an @username is provided.

File: lib/workers/repository/update/pr/participants.ts
Comment (bbckr) [reply lib/workers/repository/update/pr/participants.ts:15]: Let me know if I should proceed w/ filtering out emails from the list. I think its fine to have the list normalized here since it needs to be normalized shortly after in the same flow.

Comment (viceice) [reply lib/workers/repository/update/pr/participants.ts:15]: ok, at least filter emails so we can reduce failed. requests. I also think we should remove the `@` directly before the request

Comment (viceice) [reply lib/workers/repository/update/pr/participants.ts:15]: removing the `@` here can have negative impact on other platform.

so please do the email filtering on gitlab platform code.

File: lib/workers/repository/common.ts
Comment (viceice) [inline lib/workers/repository/common.ts:20]: move to util/common

File: lib/modules/platform/gitlab/index.ts
Comment (viceice) [inline lib/modules/platform/gitlab/index.ts:1359]:
```suggestion
    if (reviewerOrAssignee.indexOf('@') > 1) {
```
otherwise you filter all which are starting with `@` too

File: lib/workers/repository/update/pr/participants.ts
Comment (viceice) [inline lib/workers/repository/update/pr/participants.ts:25]: where is this used? is the `@` already stripped somewhere?

Comment (bbckr) [reply lib/workers/repository/update/pr/participants.ts:25]: Yes, in two places. More precisely it is called in `addParticipants` when we `prepareParticipants` if there is at least 1 or more assignees. So actually this occurs immediately after addCodeOwners (if codeowners configured).

The same logic applies to reviewers as well.

File: lib/modules/platform/gitlab/index.ts
Comment (bbckr) [reply lib/modules/platform/gitlab/index.ts]: > 0 should be correct because if it starts with an @, it would be at index 0. So I want greater than index 0 (incase we have `a@email.com`, which would be index 1.

**Comment by bbckr** [comment]
Hey @viceice and @rarkins! Just checking in. Let me know if I need to follow-up with anything else. I've requested re-reviews, and thanks again for taking yalls time to do the initial reviews 🙏

**Review by rarkins — DISMISSED** [review]

File: lib/modules/platform/gitlab/index.ts
Comment (viceice) [inline lib/modules/platform/gitlab/index.ts:1376]: do we get a 404 if the group isn't found? so we log the error if it's not 404?

File: lib/util/common.ts
Comment (viceice) [inline lib/util/common.ts:64]: @rarkins @secustor wonder why we do a length check here. it doesn't catch or improve anything here.
```suggestion
  return input.startsWith('@') ? input.slice(1) : input;
```

**Review by viceice — COMMENTED** [review]
otherwise LGTM

File: lib/util/common.ts
Comment (secustor) [reply lib/util/common.ts:64]: Is this copied from somewhere?

Comment (viceice) [reply lib/util/common.ts:64]: yes, code was moved from worker below

Comment (secustor) [reply lib/util/common.ts:64]: Then 🤷‍♂️

File: lib/modules/platform/gitlab/index.ts
Comment (bbckr) [reply lib/modules/platform/gitlab/index.ts:1376]: we do get a 404 if the group is not found, so whether that is the reason (or its auth related, etc.), we debug log for all those cases.

Comment (viceice) [reply lib/modules/platform/gitlab/index.ts:1376]: ok, we should silently ignore 404 or at least don't log the http error

File: lib/util/common.ts
Comment (viceice) [inline lib/util/common.ts:66]:
```suggestion
  return input.startsWith('@') ? input.slice(1) : input;
```
the other check makes no sense, so just drop it

File: lib/modules/platform/gitlab/index.ts
Comment (viceice) [inline lib/modules/platform/gitlab/index.ts:1374]: I told to only drop log when 404 but log all other errors

something like `if(err.statusCode !== 404) logger.debug(...);`

File: lib/util/common.ts
Comment (bbckr) [reply lib/util/common.ts]: updated and removed

File: lib/modules/platform/gitlab/index.ts
Comment (bbckr) [reply lib/modules/platform/gitlab/index.ts:1377]: updated

**Review by viceice — DISMISSED** [review]

File: lib/modules/platform/gitlab/index.ts
Comment (viceice) [inline lib/modules/platform/gitlab/index.ts:1374]: needs coverage 😉

**Review by rarkins — DISMISSED** [review]

**Review by viceice — DISMISSED** [review]

**Comment by bbckr** [comment]
@rarkins @viceice reviews got dismissed since I added one additional line to test that we log the actual message "Unable to fetch groups", will need one more review/approval

**Review by viceice — APPROVED** [review]

**Comment by rarkins** [comment]
@bbckr thanks for your first Renovate PR! Can you send me an email if you're interested in a voucher for a free Renovate t-shirt? Email address is in my profile

**Comment by bbckr** [comment]
> @bbckr thanks for your first Renovate PR! Can you send me an email if you're interested in a voucher for a free Renovate t-shirt? Email address is in my profile

Definitely interested, just sent an email. Thanks!

**Pull request merged**
